Use `xparse` commands in new commands/environments tutorial (#32)
oetiker/lshort: book/src/custom.tex

\section{New Commands, Environments and Packages}

At the beginning of this book we have mentioned that \LaTeX{} allows us to

> write documents using logical markup, such as \ci{emph} or \ci{section}. There

Reviewer: ```suggestion
write documents using logical markup, with commands like \ci{emph} or \ci{section}. There
```

may be however situations where \LaTeX{} does not provide an appropriate command for the content you want write about.```
You may have noticed that all the
commands I introduce in this book are typeset in a box, and that they show up

> in the index at the end of the book. There is no \LaTeX{} markup to format example code or commands, but \LaTeX{} allows me to define new commands. Thus I

Reviewer: ```suggestion
in the index at the end of the book. There is no \LaTeX{} markup to format example code or commands, but \LaTeX{} allows me to define my own commands for this purpose. Thus I
```

\section{New Commands, Environments and Packages}

At the beginning of this book we have mentioned that \LaTeX{} allows us to
write documents using logical markup, such as \ci{emph} or \ci{section}. There

> may be however situations where a command to produce the logical unit of your

Reviewer: ```suggestion
may be however situations where \LaTeX{} does not provide an appropriate command for the content you want write about```

> document is not provided by \LaTeX{}. You may have noticed that all the

Reviewer: ```suggestion
You may have noticed that all the
```

> commands I introduce in this book are typeset in a box, and that they show up
> in the index at the end of the book. There is no \LaTeX{} markup for element
> ``this is a command introduction'', bu \LaTeX{} allows me to define one. Thus I
> can write:

Reviewer: ```suggestion
commands I introduce in this book are typeset in a box, and that they show up
in the index at the end of the book. There is no \LaTeX{} markup to format example code or commands, but \LaTeX{} allows me to define new commands. Thus I
can write:
```

command that takes two arguments. In order to refer to the received arguments
you use \verb|#1| for the first argument, \verb|#2| for the second, and so on.

\begin{example}
\NewDocumentCommand{\txsit}{mm}
 {This is the \emph{#1}
  #2 Introduction to \LaTeXe}

% in the document body:
\begin{itemize}
\item \txsit{not so}{short}
\item \txsit{very}{long}
\end{itemize}
\end{example}

If your command accepts an optional argument but the user does not supply one,
a special marker \cargv{-NoValue-} will be inserted instead.
\begin{example}
\NewDocumentCommand{\txsit}{om}
{This is the \emph{#1}
  #2 Introduction to \LaTeXe}

% in the document body:
\begin{itemize}
\item \txsit{definitive}
\item \txsit[very]{long}
\end{itemize}
\end{example}
In order to test whether the user supplied a value, use
\begin{lscommand}
  \ci*{IfValueTF}[argument:m ! value version:m ! no value version:m] 
\end{lscommand}
macro.
\begin{example}
\NewDocumentCommand{\MyCommand}{o}{%
  \IfValueTF {#1} {%
    Optional argument: #1.%
  } {
    No optional argument given.%
  }%
}

\MyCommand\\
\MyCommand[hello]
\end{example}

There are two variations of it: \ci{IfValueT} and \ci{IfValueF} which may

> be used if you only need output for one of the branches. Thus the example with

Reviewer: ```suggestion
be used if you only need output for one of the branches. The example with
```

\cargv{-NoValue-} in the output could be fixed by writing
\begin{example}
\NewDocumentCommand{\txsit}{om}
{This is the
  \IfValueT{#1}{\emph{#1} }%
  #2 Introduction to \LaTeXe}

% in the document body:
\begin{itemize}
\item \txsit{definitive}
\item \txsit[very]{long}
\end{itemize}
\end{example}

> There are also commands \ci{IfNoValueTF}, \ci{IfNoValueT} and \ci{IfNoValueF}

Reviewer: ```suggestion
The commands \ci{IfNoValueTF}, \ci{IfNoValueT} and \ci{IfNoValueF}
```

> which work exactly the same, but the value\slash no-value branches are swapped.

Reviewer: ```suggestion
work exactly the same, but the value\slash no-value branches are swapped.
```

Often you will want to use optional arguments when present but use some default
values when the user does not provide them. This could be achieved by

> \ci{IfValueTF}, but a much better approach exists: \cargv{O} argument

Reviewer: ```suggestion
\ci{IfValueTF}, but with the \cargv{O} argument a default value can be set directly.
```

> specification. It works like \cargv{o} but allows us to specify default value

Reviewer: ```suggestion
It works like \cargv{o} but allows us to specify default value
```

> if no value is supplied. Thus we can write

Reviewer: ```suggestion
if no value is supplied. Write
```

\begin{example}
\NewDocumentCommand{\txsit}{O{not so}m}
{This is the \emph{#1}
  #2 Introduction to \LaTeXe}

% in the document body:
\begin{itemize}
\item \txsit{definitive}
\item \txsit[very]{long}
\end{itemize}
\end{example}

Another useful argument specification is \cargv{s}, short for star. This

> argument enables us to provide different definitions based on whether the

Reviewer: ```suggestion
argument allows to provide different definitions based on whether the
```

starred or non-starred version of command was issued by the user. It uses
\ci{IfBooleanTF} command (and its variations \cih{IfBooleanT} \cih{IfBooleanF})
that works analogously to the \ci{IfValueTF} command. 

\begin{example}
\NewDocumentCommand{\txsit}{sO{not so}m}
{This is the \emph{#2}
  #3 Introduction to \LaTeXe%
  \IfBooleanT{#1}{%
    : Superstar Edition%
  }%
}

% in the document body:
\begin{itemize}
\item \txsit{long}
\item \txsit*{long}
\end{itemize}
\end{example}

These are just the most common argument specifications. For a full description,
take a look at the \pai*{xparse} package documentation.\footnote{This
package is loaded automatically in new \LaTeX{} releases, so you probably do not

> need to put \texttt{\textbackslash usepackage\{xparse\}} in your preamble. }

Reviewer: perfect

\LaTeX{} will not allow you to create a new command that would
overwrite an existing one. But there is a special command in case you
explicitly want this: \ci{RenewDocumentCommand}.
It uses the same syntax as the \ci{NewDocumentCommand}
command.

In certain cases you might also want to use the \ci{ProvideDocumentCommand}
command. It works like \ci{NewDocumentCommand}, but if the command is
already defined, \LaTeXe{} will silently ignore it.

\subsection{New Environments}

> Just as with the \ci{NewDocumentCommand} command, there is a command to create
> your own environments. The \ci{NewDocumentEnvironment} command uses the

Reviewer: ```suggestion
The \ci{NewDocumentEnvironment} command lets you create your own environments. It uses the
```

following syntax:

\begin{lscommand}
\ci*{NewDocumentEnvironment}[name:m ! argspec:m ! at begin:m ! at end:m]
\end{lscommand}

The \carg{argspec} argument accepts the same arguments as the
\ci{NewDocumentCommand} command. The contents of \carg{at begin} and \carg{at
end} arguments will be inserted respectively when the commands
\ci{begin}[name] and \ci{end}[name] is encountered.

The example below illustrates the usage of the \ci{NewDocumentEnvironment}
command.
\begin{example}
\NewDocumentEnvironment{king}{}{%
  \emph{Listen! For the
    king made a statement:}%
  \\[1em]%
} {%
  \\[1em]%
  \emph{This concludes
    the king's statement.}%
}

\begin{king}
My humble subjects \ldots
\end{king}
\end{example}

> Note that when environment argument are read they are read \emph{after} the

Reviewer: ```suggestion
Note that when environment argument are read, they are read \emph{after} the
```

  }%
} {%
  \\[1em]%
  \emph{This concludes
    the king's statement.}%
}

\begin{king}*
My humble subjects \ldots
\end{king}
\end{example}

If you want to create a starred version of an environment (similarly to some
\hologo{AmSLaTeX} environments) you have to define it separately.
\begin{verbatim}
\NewDocumentEnvironment{king}{moo} { ... } { ... }
\NewDocumentEnvironment{king*}{moo} { ... } { ... }
\end{verbatim}
Obviously you can use the same internal commands to define them, for example
renaming the previous implementation to \cargv{kinginternal} making the
\cargv{king} and \cargv{king*} a thin wrapper around it.

The \ci{NewDocumentEnvironment} also introduces a special argument
specification: \cargv{+b}, short for body.\footnote{The \cargv{+} indicates
that it may contain multiple paragraphs.} It is only allowed as the last
argument in the \carg{argspec}. It allows you to receive the body of the
environment as an argument.
\begin{example}
\NewDocumentEnvironment{twice}{+b} {%
  First time:\\
  #1

  Second time:\\
  #1
} {}

\begin{twice}
This will be printed twice!
\end{twice}
\end{example}
While this makes one of the \carg{at begin}, \carg{at end} arguments redundant,

> they are still required. (In the example above we provided empty \carg{at

Reviewer: ```suggestion
they are still required. (In the example above we provided an empty \carg{at
```

typesetting.

Similarly to the \ci{NewDocumentCommand} \LaTeX{} makes sure that you do not
define an environment that already exists. If you ever want to change an
existing environment, use the \ci{RenewDocumentEnvironment} command. Its
arguments are exactly the same as the \ci{NewDocumentEnvironment} command.

\subsection{Copying commands}

When redefining commands you may want to use the original version of the
command. Your initial code may look like this
\begin{verbatim}
\RenewDocumentCommand{\emph}{m}{%
  \emph{#1}~(``#1'' is emphasised)%
}
\end{verbatim}
but when you try to compile the document you will receive the error
\begin{verbatim}
! TeX capacity exceeded, sorry [input stack size=5000].
\end{verbatim}

To understand why this happens it is instructive to consider how \TeX{} expands
the defined commands. The above \ci{RenewDocumentCommand} tells the \TeX{}
engine that whenever \ci*{emph}[foo:vm] is seen it must replace it with
\verb|\emph{foo}~(``foo'' is emphasised)|. You may already see the problem here.
In the next stage it will again replace the \ci*{emph}[foo:vm] yielding
\begin{verbatim}
\emph{foo}~(``foo'' is emphasised)~(``foo'' is emphasised)
\end{verbatim}

> It is easy to see that this process will never end and at some point \TeX{}

Reviewer: ```suggestion
This process will never end and at some point \TeX{}
```

simply gives up.

Note that 
\begin{verbatim}
\NewDocumentCommand{\oldemph}{m}{\emph{#1}}
\RenewDocumentCommand{\emph}{m}{%
  \oldemph{#1}~(``#1'' is emphasised)%
}
\end{verbatim}
will suffer the same fate since \verb|\oldemph{...}| will be replaced by \TeX{}
with \verb|\emph{...}| and the cycle continues.

In order to avoid this problem a special command exists
\begin{lscommand}
  \ci*{NewCommandCopy}[name:m ! command:m]
\end{lscommand}

> It makes the \carg{name} the exact copy of the \carg{command}. You may
> appropriate the difference between this and \ci{NewDocumentCommand} in the
> following example

Reviewer: ```suggestion
It makes the \carg{name} the exact copy of the \carg{command}. The following example shows how this works```

To add your own commands, use the
\begin{lscommand}
\ci*{NewDocumentCommand}[\texttt{\bs}name:m ! argspec:m ! definition:m]
\end{lscommand}
\noindent command. It requires three arguments: the \carg{name} of the command
you want to create, the \carg{arg}ument \carg{spec}ification and the
\emph{definition} of the command.

The \carg{argspec} argument specifies the number and types of arguments the

> command receives. Two most important types are \cargv{m} and \cargv{o} which

Reviewer: ```suggestion
command receives. The two most important types are \cargv{m} and \cargv{o} which
```

> command receives. Two most important types are \cargv{m} and \cargv{o} which
> stand for, respectively, mandatory and optional. Thus if you want to create a

Reviewer: ```suggestion
command receives. The two most important types are \cargv{m}, for \emph{mandatory} and \cargv{o} for \emph{optional}. To create a
```

Often you will want to use optional arguments when present but use some default
values when the user does not provide them. This could be achieved by
\ci{IfValueTF}, but a much better approach exists: \cargv{O} argument
specification. It works like \cargv{o} but allows us to specify default value
if no value is supplied. Thus we can write

\begin{example}
\NewDocumentCommand{\txsit}{O{not so}m}
{This is the \emph{#1}
  #2 Introduction to \LaTeXe}

% in the document body:
\begin{itemize}
\item \txsit{definitive}
\item \txsit[very]{long}
\end{itemize}
\end{example}

Another useful argument specification is \cargv{s}, short for star. This
argument enables us to provide different definitions based on whether the
starred or non-starred version of command was issued by the user. It uses
\ci{IfBooleanTF} command (and its variations \cih{IfBooleanT} \cih{IfBooleanF})
that works analogously to the \ci{IfValueTF} command. 

\begin{example}
\NewDocumentCommand{\txsit}{sO{not so}m}
{This is the \emph{#2}
  #3 Introduction to \LaTeXe%
  \IfBooleanT{#1}{%
    : Superstar Edition%
  }%
}

% in the document body:
\begin{itemize}
\item \txsit{long}
\item \txsit*{long}
\end{itemize}
\end{example}

> These are just the most common argument specifications. For a full overview of
> those take a look at the documentation of \pai*{xparse} package.\footnote{This

Reviewer: ```suggestion
These are just the most common argument specifications. For a full description of
take a look at the \pai*{xparse} package documentation.\footnote{This
```

> package is included by default in new \LaTeX{} versions, so you probably do not
> need to include it manually.}

Reviewer: ```suggestion
package is part of the current \LaTeX{} release, so you probably do not
need to install it manually.}
```

Author: This comment wasn't about installing but about doing `\usepackage{xparse}`, hence the word include. Do you think other word would be more appropriate?

Reviewer: oops :)

Reviewer: how about this wording ?

\begin{figure}[!htbp]
\begin{lined}{\textwidth}
\begin{verbatim}
\ProvidesExplPackage{demopack}{2022-05-05}{0.1}{%
  Demo Package by Tobias Oetiker
}

\NewDocumentCommand{\tnss}{} {
  The~not~so~Short~Introduction~to~\LaTeXe
}
\NewDocumentCommand{\txsit}{O{not~so}} {
  The~\emph{#1}~Short~Introduction~to~\LaTeXe
}

\NewDocumentEnvironment{king}{} {
  \begin{quote}
} {
  \end{quote}
}
\end{verbatim}
\end{lined}
\caption{Example Package.} \label{package}
\end{figure}

Writing a package basically consists of copying the contents of your document
preamble (with minor adjustments) into a separate file with a name ending in
\texttt{.sty}. There is one special command,
\begin{lscommand}
\ci*{ProvidesExplPackage}[package name:m ! date:m ! version:m ! description:m]
\end{lscommand}

> \noindent for use at the very beginning of your package file. This command
> switches the \LaTeX{} to process the  file in \emph{expl mode}. There are some
> differences in the way this mode is handled. The most visible change is the

Reviewer: ```suggestion
\noindent for use at the very beginning of your package file. This command
tells the \LaTeX{} to process the  file in \emph{expl mode}. The most visible effect of this is
```

> fact that all whitespace is ignored. You may have noticed that in a lot of

Reviewer: ```suggestion
that all whitespace is ignored. You may have noticed that in many of the
```

> examples above we had to end most lines with \ai{\%} in order to get correct

Reviewer: ```suggestion
examples above we had to end most lines with \ai{\%} to get correct
```

> spacing. This change forces us to insert spaces explicitly as they are usually

Reviewer: ```suggestion
spacing in the output. In \emph{expl mode}, spaces have to be added explicitly if needed at all. They are usually
```

> much rarer when writing a package. In order to insert spaces you have to use

Reviewer: ```suggestion
quite rare when writing a package. To insert spaces, use
```

the \ai{\~} character, which normally denotes non-breaking space.\footnote{If

>   you want to insert non-breaking space in the \emph{expl mode} you have to use

Reviewer: ```suggestion
  you want to insert non-breaking space in \emph{expl mode}, use
```

  \ci{nobreakspace}.} Paragraphs can be started with the \ci{par} command.

The arguments are used to provide information about package in the log file. If

> you used this package and looked at the log file you would find

Reviewer: ```suggestion
you use this package and look at the log file you will find
```

\begin{figure}[!htbp]
\begin{lined}{\textwidth}
\begin{verbatim}
\ProvidesExplPackage{demopack}{2022-05-05}{0.1}{%
  Demo Package by Tobias Oetiker
}

\NewDocumentCommand{\tnss}{} {
  The~not~so~Short~Introduction~to~\LaTeXe
}
\NewDocumentCommand{\txsit}{O{not~so}} {
  The~\emph{#1}~Short~Introduction~to~\LaTeXe
}

\NewDocumentEnvironment{king}{} {
  \begin{quote}
} {
  \end{quote}
}
\end{verbatim}
\end{lined}
\caption{Example Package.} \label{package}
\end{figure}

Writing a package basically consists of copying the contents of your document
preamble (with minor adjustments) into a separate file with a name ending in
\texttt{.sty}. There is one special command,
\begin{lscommand}
\ci*{ProvidesExplPackage}[package name:m ! date:m ! version:m ! description:m]
\end{lscommand}
\noindent for use at the very beginning of your package file. This command
switches the \LaTeX{} to process the  file in \emph{expl mode}. There are some
differences in the way this mode is handled. The most visible change is the

> fact that whitespaces are ignored. You may have noticed that in a lot of

Reviewer: ```suggestion
fact that all whitespace is ignored. You may have noticed that in a lot of
```
